% !TEX root = main.tex

% state the problem
% RNA secondary structure prediction is a well-known problem, and it has been used for medical design. 
% Compared with MFE-based methods, partition function-based methods attract more and more attention due to their higher accuracy and ability to predict pseudoknots.
% Recently, 
% For example, almost every week a new ncRNA is found to be regulated in a particular disease, or a new class of noncoding
% transcripts is uncovered by a transcriptomic study, or a new
% article heralds a paradigm shift that lncRNAs will bring to
% our understanding of biology 
% Each year, noncoding RNA 
% Ribonucleic acid (RNA) 

\section{Introduction}

% \begin{figure}[t]
% \center
% \begin{tabular}{cc}
% \hspace{-3cm}\panel{A} & \hspace{-3cm}\panel{B} \\
% \includegraphics[scale=0.3]{figs/gold_RNAstructure}
% &
% \includegraphics[scale=0.3]{figs/mfe_RNAstructure}\\
% \hspace{-3cm}\panel{C} & \hspace{-3cm}\panel{D}\\[-0.4cm]
% \includegraphics[scale=0.085]{figs/tRNA_circular}
% &
% \includegraphics[scale=0.4]{figs/tRNA_heatmap_dark}

% \end{tabular}
% \caption{
% Comparison of MFE-based method and partition function-based method. 
%     {\bf A}: ground truth secondary structure of {\it E.~coli} tRNA$^\textit{Gly}$; 
%     {\bf B}: the corresponding MFE structure. 
%     Structural difference are denoted with blue in ground truth structure and red in MFE structure;
%     {\bf C}: the corresponding circular representation.
%     Ground truth base pairs are denoted with dash blue lines. 
%     Base pair probabilities are denoted with red solid lines and line shade is proportional to probability value.
%     {\bf D}: the corresponding heatmap representation.
%     MFE structure (lower triangle) misses some ground truth base pairs (blue cross), 
%     while base pairing probability matrix (upper triangle) covers these correct base pairs. 
% \label{tRNA}}
% \end{figure}


% For past decades, our understanding of ribonucleic acid (RNA) is changing. 
% New proofs reveal that 
% noncoding 
% RNAs %(ncRNAs) 
% Ribonucleic acids (RNAs)
RNAs
are involved in multiple processes, 
such as catalyzing reactions or guiding RNA modifications~\cite{Eddy:2001,Doudna+Cech:2002,Bachellerie+:2002}, 
% and regulating a particular disease~\cite{Kung+:2013},
and their functionalities are highly related to structures.
% but determining the structure using experimental methods is costly and time-comsuming. 
%%%%%%%%5
% from proposal
% Therefore, being able to %rapidly 
% determine the structure is %extremely 
% useful and desired.
% given the overwhelming pace of increase in genomic data (about 1021 base-pairs per year \cite{stephens+:2015}) %[97] 
% and given the small percentage of sequences that have experimentally determined structure. 
% While experimental assays still constitute the most reliable way to determine structures, they are prohibitively costly, slow, and difficult.
However, 
% both 
structure determination techniques, such as X-ray crystallography~\cite{Zhang+Adrian:2014}, 
Nuclear Magnetic Resonance (NMR)~\cite{Zhang+Keane:2019}, 
and 
% chemical probing methods 
cryo-electron microscopy~\cite{Lyumkis:2019}, 
% ~\cite{Ziehler+Engelke:2001},
though reliable and accurate,
are extremely slow and costly.
% considering the exponentially increasing genomic data (about $10^{21}$ base-pairs per year \cite{stephens+:2015}) 
% and undetermined structures.
% and therefore computational prediction provides an attractive alternative.
%%%%%%%%%%%%
% Due to such limitations, fast and accurate RNA structure prediction is required and desired,
% Due to such limitations, 
% for many RNA tasks 
Therefore,
fast and accurate computational prediction of RNA structure is useful and desired. %required. 
Considering full RNA %tertiary 
structure prediction is challenging~\cite{Miao+:2017},
% \cite{mccaskill:1990},
% even more difficult than protein folding \cite{mccaskill:1990},
% as an alternative
many studies focus on predicting secondary structure,
% the double helices folding structure formed by self-complementary nucleotides
the set of canonical base pairs in the structure 
(A-U, G-C, G-U base pairs)~\cite{Tinoco+Bustamante:1999},
% RNA secondary structure prediction
as it is well-defined, 
% in mathematics formation, 
and provides detailed information to help understand 
% RNA's mechanism of functionality,
the structure-function relationship, and
% functionality 
% as well as further 
%The secondary structure additionally
is a basis to predict full tertiary structure~\cite{Flores+Altman:2010,Seetin+Mathews:2011}.

\input{fig_overview}



% Secondary structure prediction problem, 
% though still difficult, 
% is well-defined in mathematics formation, and can be suitable modeled with the decomposable substructures. 
% Utilizing this decomposable nature, 
RNA secondary structure prediction is NP-complete~\cite{Pedersen+:2000},
but nested (i.e., pseudoknot-free) secondary structures can be predicted with
cubic-time dynamic programming algorithms. 
% based on an important paradigm free energy minimization 
Commonly, the minimum free energy (MFE) 
structure is predicted~\cite{nussinov+jacobson:1980, zuker+stiegler:1981}.
% when a single structure is expected.
% and some prediction systems based on these algorithms, such as RNAstructure \cite{mathews+turner:2006}, \contrafold \cite{do+:2006} and \viennarnafold \cite{lorenz+:2011}, 
% have greatly improved the accuracy of prediction and are widely used.
% MEA -> partition function 
% For a given sequence, predicting the structure of minimum free energy (MFE) under certain free energy model by dynamic programming is a classical method for RNA secondary structure prediction. 
% cted.Free energy minimization is an important paradigm for predicting structure when a single structure is expe
% In the absence of many homologous sequences, the accuracy of MFE structure is 73\% in average \cite{mathews:2004}.
% However, this method assumes all thermodynamic parameters are correct and the energy model is perfect, which are both no the case in reality.
% Also, this method neglects the facts that multiple conformations exits at equilibrium \cite{mathews:2004}.
% This method 
% MFE method gives a practical solution to predict a single secondary structure, 
At equilibrium, the MFE structure is the most populated structure, 
but it 
% it neglects the fact that 
is a simplification because 
multiple conformations exist 
% at equilibrium 
as an equilibrium ensemble 
for one RNA sequence~\cite{mathews:2004}.
For example, many mRNAs {\textit {in vivo}} form a dynamic equilibrium 
and fold into a population of structures~\cite{Long+:2007, Lu+:2008, Tafer+:2008, lai+:2018}; 
% as well as abandons all pseudoknotted structures.
% Many RNA sequences, for example mRNAs, exist in a thermodynamic ensemble of structures 
% \cite{lai+:2018}.%[53].
% As an alternative, partition function-based method \cite{mccaskill:1990} provides an ensemble of all pseudoknot-free structures, and based on it base pairing probabilities and structural entropy \cite{Huynen+} can be calculated.  
% As an alternative, 
Figure~\ref{fig:overview}A--B shows the example of Tebowned RNA
which folds into more than one structure at equilibrium.
% ~\cite{Cordero+Das:2015}.
% TBWN-A, which has a long helices at 3'-end, 
% is the majority structure and accounts for $56 \pm 16\%$ of ensemble.
% While TBWN-B, which has two short helices at 5'-end, takes up $27 \pm 12\%$ of ensemble.
% Besides TBWN-A and TBWN-B illustrated in Figure~\ref{fig:overview}, Tebowned RNA can also fold into the state of TBWN-C 
% with a smaller ensemble proportion of $17 \pm 11\%$.
In this case, the prediction of one single structure, such as the MFE structure, 
is not expressive enough to capture multiple states of RNA sequences %Tebowned RNA %folding.
at equilibrium.

% Rather than predicting one single stucture, 
% partition function-based methods estimate the folding in a different point of view by
% taking into account ensemble of structures at equilibrium with Boltzmann Distribution.  
% Partition function, the sum of equilibrium constants for all possible secondary structures,
% is the normalization terms for calculating given secondary structure in the boltzmann ensemble.
Alternatively, we can compute the partition function, 
which is the sum of the equilibrium constants for all possible secondary structures,
and is the normalization term for calculating the probability of a secondary structure in the Boltzmann ensemble.
% Starting from the partition function,
% these methods are able to model mix of conformations,
% and further 
% we can % also
The partition function calculation can also be used to 
calculate base pairing probabilities of each nucleotide $i$ 
paired with each of possible nucleotides $j$~\cite{mccaskill:1990, mathews:2004}. 
% Base pairs with high probabilities 
% %in the matrix 
% indicate strong confidence of pairing in prediction,
% and are more likely to be paired in ground truth structures
% \cite{mathews:2004, Zuber+:2018}. 
% Figure~\ref{fig:overview}A 
In Figure~\ref{fig:overview}C,
the upper triangle presents the base pairing probability matrix of Tebowned RNA using \viennarnafold, 
showing that base pairs in TBWN-A have higher probabilities (in darker red) than
base pairs in TBWN-B (in lighter red).
This is consistent with the experimental result, i.e.,
TBWN-A is the majority structure that accounts for $56 \pm 16\%$ of the ensemble, 
while TBWN-B takes up $27 \pm 12\%$~\cite{Cordero+Das:2015}. % of the ensemble

% ~\cite{Cordero+Das:2015}.
% Besides producing base pairing probability matrix, 

%% \begin{table}
%% %  \vspace{-0.5cm}
%%   \centering
%%   \begin{tabular}{@{}c@{ }|@{}c@{ }|l@{\!}c@{ }|l@{\!}c@{}}
%%     & span & \multicolumn{2}{c|}{minimum free energy} & \multicolumn{2}{c}{partition-function} \\
%%     \hline
%%     bottom-up, {\em exact} &$\infty$ &    Zuker\cite{zuker+stiegler:1981} & $O(n^3)$ & McCaskill\cite{mccaskill:1990} & $O(n^3)$ \\    
%%     \hline
%%     \mtwo{local folding} & \mtwo{$L$}  &  \mtwo{Localfold\cite{lange+:2012}} & \mtwo{$O(nL^2)$} & {RNAplfold\cite{bernhart+:2006}} & \mtwo{$O(nL^2)$}\\
%%                   &           &   &       &   Rfold\cite{kiryu+:2008}        &\\
%%     \hline    
%%     left-to-right, {\em exact}  & \mtwo{$\infty$} & \mtwo{\linearfold\cite{huang+:2019}} & $O(n^3)$  & \mtwo{\linearpartition} & $O(n^3)$\\
%%      \quad + {\em beam pruning} &     && $O(n b\log b)$ && $O(n b^2)$
%%     %% left-to-right & \mtwo{\linearfold, $O(n b\log b)$} & \linearpartition, $O(n b^2)$\\[-0.1cm]
%%     %% + {\em beam pruning} & 
%%   \end{tabular}
%%   \caption{Comparison between classical, local, and left-to-right algorithms for MFE and partition function calculation. 
%%     \linearfold and \linearpartition enjoy linear runtime thanks to left-to-right order which enables heuristic beam pruning,
%%     and both become exact $O(n^3)$ algorithms without pruning. % size $b$ is $+\infty$.
%%     ``Span'' denotes the maximum pair distance allowed ($\infty$ means no limit);
%%     it is a small constant in local methods (e.g., default $L$=70 in RNAplfold).
%%     \label{tab:overall}
%%   } 

%% \end{table}


\input{fig_numpairs}

In addition to model multiple states at equilibrium, 
base pairing probabilities are used for downstream prediction methods, 
such as maximum expected accuracy (MEA)~\cite{Knudsen+Hein:2003, do+:2006}, 
to assemble a structure with improved accuracy compared with the MFE structure~\cite{lu+:2009}.
% As a by-product, the pair probabilities also enable maximum expected accuracy (MEA) structure prediction 
% \cite{do+:2006, lu+:2009}. % [29, 62].
Other downstream prediction methods, 
such as 
% HotKnot \cite{Ren+:2005}, % not based on partition function
ProbKnot~\cite{bellaousov+mathews:2010}, 
ThreshKnot~\cite{Zhang+:2019},
DotKnot~\cite{Sperschneider+Datta:2010} 
and IPknot~\cite{sato+:2011},
use base pairing probabilities to predict pseudoknotted structures with heuristics,
which is beyond the scope of standard cubic-time algorithms.
Additionally, the partition function 
% can also be applied to do stochastic sampling based on the ensemble distribution
is the basis of stochastic sampling, 
in which structures are sampled with their probability of occurring 
in the Boltzmann ensemble~\cite{Ding+Lawrence:2003, mathews:2006}.

% Moreover, although partition function-based method excluded pseudoknotted structures during dynamic programming process, 
% it is able to predict pseudoknotted base pairs and structure by using pair probability matrix, and pseudoknotted prediction systems such as HotKnot \cite{Ren+:2005}, ProbKnot \cite{bellaousov+mathews:2010}, DotKnot \cite{Sperschneider+Datta:2010} and IPknot \cite{Sato+:2011} all take pair probability matrix as inputs. 
% Figure~\ref{tRNA} compares MFE-based and partition function-based methods.
% Furthermore, single structure prediction based on partition function calculation, such as maximum expected accuracy (MEA) ThreshKnot \cite{do+:2006, threshknot}, achieves higher accuracy in average. 


% speed
Therefore, there has beenÂ a shift from the classical MFE-based methods to partition function-based ones.
These latter methods, 
as well as the prediction engines based on them, such as partition function-mode of RNAstructure~\cite{mathews+turner:2006}, 
\viennarnafold~\cite{lorenz+:2011}, 
and \contrafold~\cite{do+:2006},
are all based on the seminal algorithm that McCaskill pioneered~\cite{mccaskill:1990}.
It employs a dynamic program to capture all possible (exponentially many) nested structures,
but its $O(n^3)$ runtime still scales poorly for longer sequences. 
This slowness %of partition function-based methods
is even more severe than the $O(n^3)$-time MFE-based ones
%with the same $O(n^3)$ time complexity
due to a much larger constant factor.
For instance, for 
% {\it E.~coli} 
{\it H.~pylori} 23S rRNA (sequence length 2,968~{\it nt}),
\viennarnafold's %(version 2.4.11)
computation of the partition function and base pairing probabilities
is 9$\times$ slower than MFE (71 vs.~8 secs),
%% takes 8 seconds to find the MFE structure, 
%% but %36+37=
%% 71 seconds to compute the partition function
%% %and another 37 seconds for
%% and base pairing probabilities, %, respectively,
%% which is a 9x slowdown;
%it is even worse for
%to make things worse,
and \contrafold % is even slower,
%% which takes about 6 seconds to find the MFE structure, % prediction,
%% but 50+70=120 seconds to compute the partition function and base pairing probabilities, respectively,
%gives a
is even 20$\times$ slower (120 vs.~6~secs).
The slowness %(both $O(n^3)$-time and the constant factor)
prevents their applications to longer sequences.
% 132.248859 contrafold

% , which is more than 9 $\times$ slower. 
% because function-based method takes two-round cubic loops for inside-outside calculation.

To address this $O(n^3)$-time bottleneck, %alleviate the cubic-factor slowness, 
we present \linearpartition, 
which is inspired by our %efficient linearization idea of \linearfold \cite{huang+:2019}
recently proposed \linearfold algorithm~\cite{huang+:2019} 
that approximates the MFE structure in linear time.
Using the same idea, \linearpartition can approximate
the partition function and base pairing probability matrix in linear time.
% Recently, \linearfold \cite{huang+:2019} 
% presents the first linear-time and linear-space MFE-based RNA folding prediction system.
% For the same 
% % {\it E.~coli} 
% {\it H.~pylori} 23S rRNA, \linearfold spends about 2 seconds, leading to a 4$\times$ runtime decrease.
% % Overall, \linearfold achieves significant efficiency and scalability improvement and higher accuracy. 
% % the first linear-time MFE-based (approximate) algorithm for RNA folding, 
% % achieves significant efficiency and scalability improvement and higher accuracy 
% % than classical $O(n^3)$ MFE-based method, especially on long sequences. 
% Inspired by the efficient linearization idea of \linearfold, we present \linearpartition, 
% which can approximate the partition function and base pairing probability matrix in linear time, 
% addressing speed bottleneck in existing systems.
% Similar as \linearfold, 
Like \linearfold,
\linearpartition % incrementally 
scans the RNA sequence from 5'-to-3'
using a left-to-right dynamic program that runs in $O(n^3)$ time,
but unlike the classical bottom-up McCaskill algorithm~\cite{mccaskill:1990} with the same speed,
our left-to-right scanning makes it possible to apply the beam pruning heuristic~\cite{Huang+Sagae:2010} 
%to narrow down the search space, 
to achieve linear runtime in practice; see Fig~\ref{fig:overview}D.
 % with substructure of lower energy.
% and only retain states with top $b$ free energy of ensemble (inside score), 
% where $b$ is the beam size.
% Though introducing beam prune to filter some structures,
Although the search is approximate,
the well-designed heuristic ensures 
%that only states with worse ensemble free energy
the surviving structures capture the bulk of the free energy of the ensemble.
It is important to note that, unlike local folding methods in Fig.~\ref{fig:overview}D,
our algorithm does {\em not} impose any limit on the base-pairing distance;
in other words, it is a {\em global} partition function algorithm.
%and the resulting partition function is close to the exact version.

%  (inside score) 
% are pruned,
% and partition function is still similar as optimal algorithm.
% and the majority is catched.

More interestingly, as Figure~\ref{fig:linearpairs} shows,
even with the $O(n^3)$-time McCaskill algorithm, %(like the one implemented in \viennarnafold),
the resulting number of base pairings with reasonable probabilities (e.g., >0.001)
grows only linearly with the sequence length.
This suggests that our algorithm, which only computes $O(n)$ pairings by design,
is a reasonable approximation.

% \linearpartition, inherits efficiency and accuracy of \linearfold. 

% \linearpartition is 11$\times$ faster than \viennarnafold for the longest sequence ({\it H.~pylori} 23S rRNA, 2,968 nucleotides) in the ArchiveII dataset,
% and 256$\times$ faster for the longest sampled sequence (15,780 nucleotides) from RNAcentral that \viennarnafold can run.
\linearpartition is 2,771$\times$ faster than \contrafold for the longest sequence (32,753~{\it nt})
% 32,767 nucleotides) 
that \contrafold can run 
in the dataset (2.5 days vs.~1.3 min.).
% Meanwhile, \linearpartition leads to a small improvement on PPV and Sensitivity compared with \viennarnafold.
% % in both MEA and ThreshKnot prediction 
% % using the probability matrix computed in linear time.
% Surprisingly, \linearpartition achieves higher accuracy improvement on longer families (16S and 23S rRNA).
Interestingly, \linearpartition is orders of magnitude faster {\em without} sacrificing accuracy.
In fact, the resulting base pairing probabilities are even better correlated with ground truth structures,
and 
when applied to downstream structure prediction tasks,
%such as MEA and ThreshKnot (a thresholded version of ProbKnot),
they lead to a small accuracy improvement on longer families (small and large subunit rRNA),
as well as a substantial accuracy improvement on long-distance base pairs (500+ \nts apart).
%% Our work is the first major speedup on this important problem in 30 years, and
%% enables calculations on full length sequences such as mRNAs.
%% has numerous applications
%% such as  sampling.

% \begin{itemize}
% \item Present an alternative left-to-right dynamic programming fashion for partition function calculation.
% \item The first algorithm to achieve linear runtime and space for partition function and base pairing probability calculation.
% % \item . 
% \end{itemize}

% more striking than LinearFold
Although both \linearpartition and \linearfold use linear-time beam search,
the success of the former is arguably more surprising,
since rather than finding one single optimal structure, 
\linearpartition needs to sum up exponentially many structures
that capture the bulk part of the ensemble free energy.
% For users, 
\linearpartition also results in more accurate downstream structure predictions than \linearfold.
%and is able to serve tasks such as pseudoknot prediction and accessibility calculation.



